### PR TITLE
Backend/profile endpoints 57

### PR DESCRIPTION
**Pull request by recombinationBTRDRS**
[Not DONE]
Backend: GET /api/profiles/{id}/ and PUT/PATCH /api/profiles/{id}/ (profile read & update) #57
https://github.com/ITA-Dnipro/UA-4421/issues/57
Знімок екрана 2026-02-03 080531

### What was done

- Added user profile fields (slug, short description, about_html, contact info, media urls, visibility)
- Implemented user tags via M2M with explicit through model
- Introduced unique and URL-safe user slug (defaulted from username, editable later)
- Fixed migrations graph inconsistency by adding missing migrations for projects and startups
- Updated users migration dependencies to match actual database state
- Resolve startups migration conflict (identical migrations)
 ### Why

- Enables public user profile pages with clean URLs
- Ensures slug uniqueness and future editability
- Prevents migration graph errors caused by missing applied migrations
- Syncs repository state with already-applied database schema

 ### Result

- Migration graph is consistent
- No fake migrations or DB resets required
- Project is ready for further profile-related features

### After first review

- Refactored slug handling to rely on default UUID generation at creation time and explicit validation on update.
- Improved profile update responses to match the public profile representation and maintain API contract consistency.

### After second review

- Slug generation is now handled during user creation, with logic moved into serializers and views. 
- Introduced CustomUserManager to centralize user creation, fully compatible with Django’s auth flow. 
- Fixed the issue where profile owners couldn’t access their own hidden profiles. 
- Refactored profile tests to cover all key scenarios without unnecessary duplication. 
- Minor improvements: stronger URL validation, better PUT/PATCH handling, improved slug validation, and cleaner code readability.
- Fixed profile serializer PATCH behavior by adjusting required fields. 
- Restructured profile API tests for clarity, added edge cases, and covered previously untested paths. 
- Added a dedicated permission test suite for IsOwnerOrReadOnly, ensuring full coverage and separation from API tests.

Знімок екрана 2026-02-08 080320

Знімок екрана 2026-02-08 080532

### CODE REVIEW

File: startup_gateway/users/models.py [hunk 1]
```diff
@@ -25,20 +26,59 @@ class User(AbstractUser):
...
+        Slug rules:
+        - slug is always non-empty
+        - slug is unique
+        - on create: slug = username
+        - slug can be changed later via API (PUT / PATCH)
+        """
+        if self._state.adding and not self.slug:
+            self.slug = self.username
+
+        super().save(*args, **kwargs)
```
Comment (VovaD199) [inline startup_gateway/users/models.py]: **Problem**

- username does not guarantee uniqueness in slug space
- when creating two users with the same username → IntegrityError
- slugify is imported, but not used

**How to correctly**

- slugify(username)
- in case of conflict — add suffix (-1, -2, uuid, etc.)
- or generate slug in serializer/service, not in model.save()

File: startup_gateway/users/views.py [hunk 2]
```diff
@@ -140,5 +176,66 @@ def post(self, request):
...
+
+        serializer = ProfileUpdateSerializer(
+            user,
+            data=request.data,
+            partial=False,
+        )
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+
+        return Response(serializer.data, status=status.HTTP_200_OK)
```
Comment (VovaD199) [inline startup_gateway/users/views.py]: **Problem**

- ProfileUpdateSerializer ≠ PublicProfileSerializer
- API contract says: “Return updated resource in response”
- now the response does not match GET

**Consequences**

- client receives different JSON structure
- API consistency violation

**How to fix**
After serializer.save():

    return Response(
        PublicProfileSerializer(user).data,
        status=status.HTTP_200_OK
    )

**Review by VovaD199 — CHANGES_REQUESTED** [review]

**Review by yuliaklimchuk — APPROVED** [review]

File: startup_gateway/users/models.py (hunk 1 above)
Comment (recombinationBTRDRS) [reply startup_gateway/users/models.py]: In the User models, the slug is now generated by default using `uuid4`.  
This guarantees uniqueness and removes the need to provide it when creating a record.  

The `save` method has been removed from the models.  

In the serializer, the slug validation logic was updated,  
since we decided to keep the option to change it through `PUT` and `PATCH` methods.
image
Знімок екрана 2026-02-04 210541

File: startup_gateway/users/views.py (hunk 2 above)
Comment (recombinationBTRDRS) [reply startup_gateway/users/views.py]: PATCH and PUT endpoints now return the updated profile using PublicProfileSerializer to keep response format consistent with GET.
Знімок екрана 2026-02-04 211456
Знімок екрана 2026-02-04 211659

File: startup_gateway/users/models.py [hunk 3]
```diff
@@ -25,17 +25,43 @@ class User(AbstractUser):
...
+        default=""
+    )
     created_at = models.DateTimeField(auto_now_add=True)
 
+    # --- profile ---
+    slug = models.SlugField(
+        max_length=160,
+        default=uuid.uuid4,
+        unique=True
+    )
```
Comment (VovaD199) [inline startup_gateway/users/models.py:67]: It is better to generate slug in the serializer / service, rather than through default.

File: startup_gateway/users/views.py [hunk 4]
```diff
@@ -140,9 +167,78 @@ def post(self, request):
...
+        ]
+
+    def get_object(self, id):
+        return get_object_or_404(User, id=id)
+
+    def get(self, request, id):
+        user = self.get_object(id)
+
+        if not user.visibility:
+            return Response(status=status.HTTP_404_NOT_FOUND)
```
Comment (VovaD199) [inline startup_gateway/users/views.py:201]: **Problem**

- owner cannot view their own hidden profile
- this directly violates the UX and logic of the profile

should be

    if not user.visibility and request.user != user:
        return Response(status=404)

File: startup_gateway/tests/test_profile_api.py
Comment (VovaD199) [inline startup_gateway/tests/test_profile_api.py:1]: **Tests do not cover key acceptance cases**
There is:

- GET public
- PATCH owner
- PATCH non-owner

 **There is no:**

- PUT success
- PUT missing required fields → 400
- GET hidden profile:

- - as owner → 200
- - as anon → 404

- invalid slug → 400
- duplicate slug → 400

This **does not formally meet the acceptance criteria** (“Tests for invalid data”).

File: startup_gateway/users/serializers.py [hunk 5]
```diff
@@ -61,6 +70,154 @@ def validate(self, attrs):
...
+
+        if self.instance and self.instance.slug == value:
+            return value
+
+        if User.objects.filter(slug=value).exists():
+            raise serializers.ValidationError("This slug is already in use.")
+
+        return value
+
+    # ---------------- MEDIA / CONTACT ----------------
```
Comment (VovaD199) [inline startup_gateway/users/serializers.py:159]: WHAT is this and why is it here?

File: startup_gateway/users/serializers.py [hunk 6]
```diff
@@ -61,6 +70,154 @@ def validate(self, attrs):
...
+        if not isinstance(value, list):
+            raise serializers.ValidationError("media_urls must be a list.")
+
+        for url in value:
+            parsed = urlparse(url)
+            if not parsed.scheme or not parsed.netloc:
+                raise serializers.ValidationError(
+                    f"Invalid URL in media_urls: {url}"
+                )
+        return value
```
Comment (VovaD199) [inline startup_gateway/users/serializers.py:180]: **URL validation in media_urls is insufficient**
Problem

- accepts ftp://, file://, javascript:
- this is an API-level security issue

Minimal fix

    if parsed.scheme not in {"http", "https"}:
        raise serializers.ValidationError("Only http/https URLs are allowed.")

**Review by VovaD199 — CHANGES_REQUESTED** [review]

File: startup_gateway/users/models.py (hunk 3 above)
Comment (recombinationBTRDRS) [reply startup_gateway/users/models.py:67]: - Slug generation now happens in the service layer.
- Validation responsibility remains in the serializer.
Знімок екрана 2026-02-08 080853

File: startup_gateway/users/views.py (hunk 4 above)
Comment (recombinationBTRDRS) [reply startup_gateway/users/views.py:201]: Done
image

File: startup_gateway/users/serializers.py (hunk 5 above)
Comment (recombinationBTRDRS) [reply startup_gateway/users/serializers.py:159]: Done 
image

File: startup_gateway/users/serializers.py (hunk 6 above)
Comment (recombinationBTRDRS) [reply startup_gateway/users/serializers.py:180]: Done 
image

File: startup_gateway/tests/test_profile_api.py
Comment (recombinationBTRDRS) [reply startup_gateway/tests/test_profile_api.py:1]: Done
image

**Review by VovaD199 — APPROVED** [review]
Once the conflicts are fixed, you can merge.

**Comment by VovaD199** [comment]
Freezing
